nucypher/config/base.py

```diff
@@ -0,0 +1,139 @@
+import json
+import os
+from abc import ABC, abstractmethod
+
+from nucypher.config import constants
+
+
+class BaseConfiguration(ABC):
+
+    _NAME = NotImplemented
+    _CONFIG_FILE_EXTENSION = 'json'
+
+    DEFAULT_CONFIG_ROOT = constants.DEFAULT_CONFIG_ROOT
+
+    class ConfigurationError(RuntimeError):
+        pass
+
+    class InvalidConfiguration(ConfigurationError):
+        pass
+
+    class NoConfigurationRoot(InvalidConfiguration):
+        pass
+
+    def __init__(self,
+                 config_root: str = None,
+                 filepath: str = None,
+                 *args, **kwargs):
+
+        if self._NAME is NotImplemented:
+            error = f'_NAME must be implemented on BaseConfiguration subclass {self.__class__.__name__}'
+            raise TypeError(error)
+
+        self.config_root = config_root or self.DEFAULT_CONFIG_ROOT
+        if not filepath:
+            filepath = os.path.join(self.config_root, self.generate_filename())
+        self.filepath = filepath
+
+        super().__init__()
+
+    def __eq__(self, other):
+        return bool(self.static_payload() == other.static_payload())
+
+    @abstractmethod
+    def static_payload(self) -> dict:
+        """
+        Return a dictionary of JSON serializable configuration key/value pairs
+        matching the input specification of this classes __init__.
+
+        Recommended subclass implementations:
+
+        ```
+        payload = dict(**super().static_payload(), key=value)
+        return payload
+        ```
+
+        OR
+
+        ```
+        subclass_payload = {'key': 'value'}
+        payload = {**super().static_payload(), **subclass_payload}
+        return payload
+        ```
+
+        """
+        payload = dict(config_root=self.config_root)
+        return payload
+
+    @classmethod
+    def generate_filename(cls, modifier: str = None) -> str:
+        name = cls._NAME.lower()
+        if modifier:
+            name += f'-{modifier}'
+        filename = f'{name}.{cls._CONFIG_FILE_EXTENSION.lower()}'
+        return filename
+
+    @classmethod
+    def default_filepath(cls):
+        filename = cls.generate_filename()
+        default_path = os.path.join(cls.DEFAULT_CONFIG_ROOT, filename)
+        return default_path
+
+    def generate_filepath(self, filepath: str = None, modifier: str = None, override: bool = False) -> str:
+        if not filepath:
+            filename = self.generate_filename()
+            filepath = os.path.join(self.config_root, filename)
+        if os.path.exists(filepath) and not override:
```

Comment:
Wait, so if the file didn't exist, we ignore the modifier? Is it how it is supposed to be by design?

Reply:
Yeah, exactly. :+1:

Reply:
If it is, I'd put that in the docstring: not obvious otherwise

Reply:
:heavy_check_mark:  - Fixed in 8e9e08a0cf4bda13e7a4428e846da5dacb7bbe14